File: LocalPayment/src/main/java/com/braintreepayments/api/localpayment/LocalPaymentClient.java
package com.braintreepayments.api.localpayment;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;

import com.braintreepayments.api.BrowserSwitchFinalResult;
import com.braintreepayments.api.BrowserSwitchOptions;
import com.braintreepayments.api.core.AnalyticsEventParams;
import com.braintreepayments.api.core.BraintreeClient;
import com.braintreepayments.api.core.BraintreeException;
import com.braintreepayments.api.core.BraintreeRequestCodes;
import com.braintreepayments.api.core.ConfigurationException;
import com.braintreepayments.api.datacollector.DataCollector;
import com.braintreepayments.api.sharedutils.Json;

import org.json.JSONException;
import org.json.JSONObject;

/**
 * Used to integrate with local payments.
 */
public class LocalPaymentClient {

    static final String LOCAL_PAYMENT_CANCEL = "local-payment-cancel";
    static final String LOCAL_PAYMENT_SUCCESS = "local-payment-success";

    private final BraintreeClient braintreeClient;
    private final DataCollector dataCollector;
    private final LocalPaymentApi localPaymentApi;

    /**
     * Used for linking events from the client to server side request
     * In the Local Payment flow this will be a Payment Token/Order ID
     */
    private String payPalContextId = null;

    /**
     * Initializes a new {@link LocalPaymentClient} instance
     *
     * @param context an Android Context
     * @param authorization a Tokenization Key or Client Token used to authenticate
     * @param returnUrlScheme a custom return url to use for browser and app switching
     */
    public LocalPaymentClient(
        @NonNull Context context,
        @NonNull String authorization,
        @Nullable String returnUrlScheme
    ) {
        this(new BraintreeClient(context, authorization, returnUrlScheme));
    }

   @VisibleForTesting
    LocalPaymentClient(@NonNull BraintreeClient braintreeClient) {
        this(braintreeClient, new DataCollector(braintreeClient),
                new LocalPaymentApi(braintreeClient));
    }

    @VisibleForTesting
    LocalPaymentClient(@NonNull BraintreeClient braintreeClient,
                       @NonNull DataCollector dataCollector,
                       @NonNull LocalPaymentApi localPaymentApi) {
        this.braintreeClient = braintreeClient;
        this.dataCollector = dataCollector;
        this.localPaymentApi = localPaymentApi;
    }

    /**
     * Starts the payment flow for a {@link LocalPaymentRequest} and calls back a
     * {@link LocalPaymentAuthRequestParams} on success that should be used to launch the user
     * authentication flow.
     *
     * @param request  {@link LocalPaymentRequest} with the payment details.
     * @param callback {@link LocalPaymentAuthCallback}
     */
    public void createPaymentAuthRequest(@NonNull final LocalPaymentRequest request,
                                         @NonNull final LocalPaymentAuthCallback callback) {
        braintreeClient.sendAnalyticsEvent(LocalPaymentAnalytics.PAYMENT_STARTED);

        Exception exception = null;

        //noinspection ConstantConditions
        if (callback == null) {
            throw new RuntimeException("A LocalPaymentAuthRequestCallback is required.");
        }

        //noinspection ConstantConditions
        if (request == null) {
            exception = new BraintreeException("A LocalPaymentRequest is required.");
        } else if (request.getPaymentType() == null || request.getAmount() == null) {
            exception = new BraintreeException(
                    "LocalPaymentRequest is invalid, paymentType and amount are required.");
        }

        if (exception != null) {
            authRequestFailure(exception, callback);
        } else {
            braintreeClient.getConfiguration((configuration, error) -> {
                if (configuration != null) {
                    if (!configuration.isPayPalEnabled()) {
                        authRequestFailure(
                            new ConfigurationException("Local payments are not enabled for this " +
                                "merchant."),
                            callback
                        );
                        return;
                    }

                    localPaymentApi.createPaymentMethod(request,
                            (localPaymentResult, createPaymentMethodError) -> {
                                if (localPaymentResult != null) {
                                    String pairingId = localPaymentResult.getPaymentId();
                                    if (pairingId != null && !pairingId.isEmpty()) {
                                        payPalContextId = pairingId;
                                    }
                                    buildBrowserSwitchOptions(localPaymentResult, request.hasUserLocationConsent(), callback);
                                } else if (createPaymentMethodError != null) {
                                    authRequestFailure(
                                        new BraintreeException("An error occurred creating the " +
                                            "local payment method."),
                                        callback
                                    );
                                }
                            });
                } else if (error != null) {
                    authRequestFailure(error, callback);
                }
            });
        }
    }

    void buildBrowserSwitchOptions(
            @NonNull LocalPaymentAuthRequestParams localPaymentAuthRequestParams,
            @NonNull boolean hasUserLocationConsent,
            @NonNull LocalPaymentAuthCallback callback
    ) {
        BrowserSwitchOptions browserSwitchOptions = new BrowserSwitchOptions()
                .requestCode(BraintreeRequestCodes.LOCAL_PAYMENT.getCode())
                .returnUrlScheme(braintreeClient.getReturnUrlScheme())
                .launchAsNewTask(braintreeClient.launchesBrowserSwitchAsNewTask())
                .url(Uri.parse(localPaymentAuthRequestParams.getApprovalUrl()));

        try {
            browserSwitchOptions.metadata(new JSONObject()
                .put("merchant-account-id",
                    localPaymentAuthRequestParams.getRequest().getMerchantAccountId())
                .put("payment-type", localPaymentAuthRequestParams.getRequest().getPaymentType())
                .put("has-user-location-consent", hasUserLocationConsent)
            );
        } catch (JSONException e) {
            authRequestFailure(
                new BraintreeException("Error parsing local payment request"),
                callback
            );
            return;
        }

        localPaymentAuthRequestParams.setBrowserSwitchOptions(browserSwitchOptions);
        callback.onLocalPaymentAuthRequest(new LocalPaymentAuthRequest.ReadyToLaunch(localPaymentAuthRequestParams));
        sendAnalyticsEvent(LocalPaymentAnalytics.BROWSER_SWITCH_SUCCEEDED);
    }

    private void authRequestFailure(Exception error, LocalPaymentAuthCallback callback) {
        sendAnalyticsEvent(LocalPaymentAnalytics.PAYMENT_FAILED);
        callback.onLocalPaymentAuthRequest(new LocalPaymentAuthRequest.Failure(error));
    }

    /**
     * After receiving a result from the web authentication flow via
     * {@link LocalPaymentLauncher#handleReturnToAppFromBrowser(LocalPaymentPendingRequest.Started, Intent)}, pass the
     * {@link LocalPaymentAuthResult.Success} returned to this method to tokenize the local
     * payment method and receive a {@link LocalPaymentNonce} on success.
     *
     * @param context                         Android Context
     * @param localPaymentAuthResult a {@link LocalPaymentAuthResult.Success} received from
     *                              {@link LocalPaymentLauncher#handleReturnToAppFromBrowser(LocalPaymentPendingRequest.Started, Intent)}
     * @param callback                        {@link LocalPaymentInternalTokenizeCallback}
     */
    public void tokenize(@NonNull final Context context,
                         @NonNull LocalPaymentAuthResult.Success localPaymentAuthResult,
                         @NonNull final LocalPaymentTokenizeCallback callback) {

        BrowserSwitchFinalResult.Success browserSwitchResult = localPaymentAuthResult
            .getPaymentAuthInfo().getBrowserSwitchResultInfo();

        JSONObject metadata = browserSwitchResult.getRequestMetadata();
        final String merchantAccountId = Json.optString(metadata, "merchant-account-id", null);
        final boolean hasUserLocationConsent = Json.optBoolean(metadata, "has-user-location-consent", false);

        Uri deepLinkUri = browserSwitchResult.getReturnUrl();
        if (deepLinkUri == null) {
            tokenizeFailure(
                new BraintreeException("LocalPayment encountered an error, return URL is " +
                    "invalid."),
                callback
            );
            return;
        }

        final String responseString = deepLinkUri.toString();
        if (responseString.toLowerCase().contains(LOCAL_PAYMENT_CANCEL.toLowerCase())) {
            callbackCancel(callback);
            return;
        }
        braintreeClient.getConfiguration((configuration, error) -> {
            if (configuration != null) {
                localPaymentApi.tokenize(merchantAccountId, responseString,
                        dataCollector.getClientMetadataId(context, configuration, hasUserLocationConsent),
                        (localPaymentNonce, localPaymentError) -> {
                            if (localPaymentNonce != null) {
                                sendAnalyticsEvent(LocalPaymentAnalytics.PAYMENT_SUCCEEDED);
                                callback.onLocalPaymentResult(new LocalPaymentResult.Success(localPaymentNonce));
                            } else if (localPaymentError != null) {
                                tokenizeFailure(localPaymentError, callback);
                            }
                        });
            } else if (error != null) {
                tokenizeFailure(error, callback);
            }
        });
    }

    private void callbackCancel(LocalPaymentTokenizeCallback callback){
        sendAnalyticsEvent(LocalPaymentAnalytics.PAYMENT_CANCELED);
        callback.onLocalPaymentResult(LocalPaymentResult.Cancel.INSTANCE);
    }
    private void tokenizeFailure(Exception error, LocalPaymentTokenizeCallback callback) {
        sendAnalyticsEvent(LocalPaymentAnalytics.PAYMENT_FAILED);
        callback.onLocalPaymentResult(new LocalPaymentResult.Failure(error));
    }

    private void sendAnalyticsEvent(String eventName) {
        AnalyticsEventParams eventParameters = new AnalyticsEventParams();
        eventParameters.setPayPalContextId(payPalContextId);
        braintreeClient.sendAnalyticsEvent(eventName, eventParameters);
    }
}

File: LocalPayment/src/main/java/com/braintreepayments/api/localpayment/LocalPaymentClient.kt
package com.braintreepayments.api.localpayment

import android.content.Context
import android.net.Uri
import androidx.annotation.RestrictTo
import com.braintreepayments.api.BrowserSwitchFinalResult
import com.braintreepayments.api.BrowserSwitchOptions
import com.braintreepayments.api.LaunchType
import com.braintreepayments.api.core.AnalyticsEventParams
import com.braintreepayments.api.core.AnalyticsParamRepository
import com.braintreepayments.api.core.BraintreeClient
import com.braintreepayments.api.core.BraintreeException
import com.braintreepayments.api.core.BraintreeRequestCodes
import com.braintreepayments.api.core.ConfigurationException
import com.braintreepayments.api.datacollector.DataCollector
import com.braintreepayments.api.sharedutils.Json
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.json.JSONException
import org.json.JSONObject
import java.util.Locale
import kotlin.coroutines.cancellation.CancellationException

/**
 * Used to integrate with local payments.
 */
class LocalPaymentClient internal constructor(
    private val braintreeClient: BraintreeClient,
    private val dataCollector: DataCollector = DataCollector(braintreeClient),
    private val localPaymentApi: LocalPaymentApi = LocalPaymentApi(braintreeClient),
    private val analyticsParamRepository: AnalyticsParamRepository = AnalyticsParamRepository.instance,
    private val dispatcher: CoroutineDispatcher = Dispatchers.Main,
    private val coroutineScope: CoroutineScope = CoroutineScope(dispatcher)
) {
    /**
     * Used for linking events from the client to server side request
     * In the Local Payment flow this will be a Payment Token/Order ID
     */
    private var contextId: String? = null

    /**
     * Initializes a new [LocalPaymentClient] instance
     *
     * @param context an Android Context
     * @param authorization a Tokenization Key or Client Token used to authenticate
     * @param returnUrlScheme a custom return url to use for browser and app switching
     */
    constructor(
        context: Context,
        authorization: String,
        returnUrlScheme: String?
    ) : this(BraintreeClient(context, authorization, returnUrlScheme))

    /**
     * Starts the payment flow for a [LocalPaymentRequest] and calls back a
     * [LocalPaymentAuthRequestParams] on success that should be used to launch the user
     * authentication flow.
     *
     * @param request  [LocalPaymentRequest] with the payment details.
     * @param callback [LocalPaymentAuthCallback]
     */
    fun createPaymentAuthRequest(
        request: LocalPaymentRequest,
        callback: LocalPaymentAuthCallback
    ) {
        coroutineScope.launch {
            val result = createPaymentAuthRequest(request)
            callback.onLocalPaymentAuthRequest(result)
        }
    }

    private suspend fun createPaymentAuthRequest(
        request: LocalPaymentRequest,
    ): LocalPaymentAuthRequest {
        analyticsParamRepository.reset()
        braintreeClient.sendAnalyticsEvent(LocalPaymentAnalytics.PAYMENT_STARTED)

        if (request.paymentType == null || request.amount == null) {
            return authRequestFailure(
                BraintreeException(
                    "LocalPaymentRequest is invalid, paymentType and amount are required."
                )
            )
        }

        return try {
            val configuration = braintreeClient.getConfiguration()
            if (!configuration.isPayPalEnabled) {
                return authRequestFailure(
                    ConfigurationException("Local payments are not enabled for this merchant.")
                )
            }

            try {
                val localPaymentResult = localPaymentApi.createPaymentMethod(request)
                val paymentId = localPaymentResult.paymentId
                if (paymentId.isNotEmpty()) {
                    contextId = paymentId
                }
                buildBrowserSwitchOptions(localPaymentResult, request.hasUserLocationConsent)
            } catch (e: Exception) {
                if (e is CancellationException) throw e
                val errorMessage =
                    "An error occurred creating the local payment method: " + e.message
                authRequestFailure(BraintreeException(errorMessage))
            }
        } catch (e: Exception) {
            if (e is CancellationException) throw e
            authRequestFailure(e)
        }
    }

    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    fun buildBrowserSwitchOptions(
        localPaymentAuthRequestParams: LocalPaymentAuthRequestParams,
        hasUserLocationConsent: Boolean,
    ): LocalPaymentAuthRequest {
        val browserSwitchOptions = BrowserSwitchOptions()
            .requestCode(BraintreeRequestCodes.LOCAL_PAYMENT.code)
            .returnUrlScheme(braintreeClient.getReturnUrlScheme())
            .launchType(LaunchType.ACTIVITY_CLEAR_TOP)
            .url(Uri.parse(localPaymentAuthRequestParams.approvalUrl))

        try {
            browserSwitchOptions.metadata(
                JSONObject()
                    .put(
                        "merchant-account-id",
                        localPaymentAuthRequestParams.request.merchantAccountId
                    )
                    .put("payment-type", localPaymentAuthRequestParams.request.paymentType)
                    .put("has-user-location-consent", hasUserLocationConsent)
            )
        } catch (e: JSONException) {
            return authRequestFailure(
                BraintreeException("Error parsing local payment request")
            )
        }

        localPaymentAuthRequestParams.browserSwitchOptions = browserSwitchOptions
        sendAnalyticsEvent(LocalPaymentAnalytics.BROWSER_SWITCH_SUCCEEDED)
        return LocalPaymentAuthRequest.ReadyToLaunch(localPaymentAuthRequestParams)
    }

    private fun authRequestFailure(error: Exception): LocalPaymentAuthRequest.Failure {
        sendAnalyticsEvent(
            eventName = LocalPaymentAnalytics.PAYMENT_FAILED,
            errorDescription = error.message
        )
        return LocalPaymentAuthRequest.Failure(error)
    }

    /**
     * After receiving a result from the web authentication flow via
     * [LocalPaymentLauncher.handleReturnToApp], pass the
     * [LocalPaymentAuthResult.Success] returned to this method to tokenize the local
     * payment method and receive a [LocalPaymentNonce] on success.
     *
     * @param context Android Context
     * @param localPaymentAuthResult a [LocalPaymentAuthResult.Success] received from
     * [LocalPaymentLauncher.handleReturnToApp]
     * @param callback [LocalPaymentInternalTokenizeCallback]
     */
    fun tokenize(
        context: Context,
        localPaymentAuthResult: LocalPaymentAuthResult.Success,
        callback: LocalPaymentTokenizeCallback
    ) {
        coroutineScope.launch {
            val result = tokenize(context, localPaymentAuthResult)
            callback.onLocalPaymentResult(result)
        }
    }

    private suspend fun tokenize(
        context: Context,
        localPaymentAuthResult: LocalPaymentAuthResult.Success,
    ): LocalPaymentResult {
        val browserSwitchResult: BrowserSwitchFinalResult.Success = localPaymentAuthResult
            .browserSwitchSuccess

        val metadata: JSONObject? = browserSwitchResult.requestMetadata
        val merchantAccountId = Json.optString(metadata, "merchant-account-id", null)
        val hasUserLocationConsent =
            Json.optBoolean(metadata, "has-user-location-consent", false)

        val deepLinkUri: Uri = browserSwitchResult.returnUrl
        val responseString = deepLinkUri.toString()
        if (responseString.lowercase(Locale.getDefault()).contains(
                LOCAL_PAYMENT_CANCEL.lowercase(Locale.getDefault())
            )
        ) {
            sendAnalyticsEvent(LocalPaymentAnalytics.PAYMENT_CANCELED)
            return LocalPaymentResult.Cancel
        }

        return try {
            val configuration = braintreeClient.getConfiguration()
            val clientMetadataID = dataCollector.getClientMetadataId(
                context,
                configuration,
                hasUserLocationConsent
            )
            val localPaymentNonce = localPaymentApi.tokenize(
                merchantAccountId,
                responseString,
                clientMetadataID
            )
            sendAnalyticsEvent(LocalPaymentAnalytics.PAYMENT_SUCCEEDED)
            LocalPaymentResult.Success(localPaymentNonce)
        } catch (e: Exception) {
            if (e is CancellationException) throw e
            tokenizeFailure(e)
        }
    }

    private fun tokenizeFailure(error: Exception): LocalPaymentResult.Failure {
        sendAnalyticsEvent(LocalPaymentAnalytics.PAYMENT_FAILED, errorDescription = error.message)
        return LocalPaymentResult.Failure(error)
    }

    private fun sendAnalyticsEvent(eventName: String, errorDescription: String? = null) {
        val eventParameters = AnalyticsEventParams(
            contextId = contextId,
            errorDescription = errorDescription
        )
        braintreeClient.sendAnalyticsEvent(eventName, eventParameters)
    }

    companion object {
        internal const val LOCAL_PAYMENT_CANCEL: String = "local-payment-cancel"
        internal const val LOCAL_PAYMENT_SUCCESS: String = "local-payment-success"
    }
}
